zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,1046 @@
+        }
+
+        // Build the actual payment values from the index.
+        let mut payments = vec![];
+        for (i, params) in params_by_index {
+            let payment = parse::to_payment(params, i)?;
+            payments.push(payment);
+        }
+
+        Ok(Zip321Request { payments })
+    }
+}
+
+pub mod render {
+    use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
+
+    use zcash_primitives::{
+        consensus, transaction::components::amount::COIN, transaction::components::Amount,
+    };
+
+    use super::{memo_to_base64, Address, RawMemo};
+
+    // The set of ASCII characters that must be percent-encoded according
+    // to the definition of ZIP 321. This is the complement of the subset of
+    // ASCII characters defined by `qchar`
+    //
+    //       unreserved      = ALPHA / DIGIT / "-" / "." / "_" / "~"
+    //       allowed-delims  = "!" / "$" / "'" / "(" / ")" / "*" / "+" / "," / ";"
+    //       qchar           = unreserved / pct-encoded / allowed-delims / ":" / "@"
+    pub const QCHAR_ENCODE: &AsciiSet = &CONTROLS
+        .add(b' ')
+        .add(b'"')
+        .add(b'#')
+        .add(b'%')
+        .add(b'&')
+        .add(b'/')
+        .add(b'<')
+        .add(b'=')
+        .add(b'>')
+        .add(b'?')
+        .add(b'[')
+        .add(b'\\')
+        .add(b']')
+        .add(b'^')
+        .add(b'`')
+        .add(b'{')
+        .add(b'|')
+        .add(b'}');
+
+    pub fn param_index(idx: Option<usize>) -> String {
+        match idx {
+            Some(i) if i > 0 => format!(".{}", i),
+            _otherwise => "".to_string(),
+        }
+    }
+
+    pub fn addr_param<P: consensus::Parameters>(
+        params: &P,
+        addr: &Address,
+        idx: Option<usize>,
+    ) -> String {
+        format!("address{}={}", param_index(idx), addr.render(params),)
+    }
+
+    pub fn amount_str(amount: Amount) -> Option<String> {
+        if amount.is_positive() {
+            let coins = i64::from(amount) / COIN;
+            let zats = i64::from(amount) % COIN;
+            Some(if zats == 0 {
+                format!("{}", coins)
+            } else {
+                format!("{}.{:0>8}", coins, zats)
+                    .trim_end_matches('0')
+                    .to_string()
+            })
+        } else {
+            None
+        }
+    }
+
+    pub fn amount_param(amount: Amount, idx: Option<usize>) -> Option<String> {
+        amount_str(amount).map(|s| format!("amount{}={}", param_index(idx), s))
+    }
+
+    pub fn memo_param(value: &RawMemo, idx: Option<usize>) -> String {
+        format!("{}{}={}", "memo", param_index(idx), memo_to_base64(value),)
+    }
+
+    pub fn str_param(label: &str, value: &str, idx: Option<usize>) -> String {
+        format!(
+            "{}{}={}",
+            label,
+            param_index(idx),
+            utf8_percent_encode(value, QCHAR_ENCODE)
+        )
+    }
+}
+
+pub mod parse {
+    use core::fmt::Debug;
+
+    use nom::{
```

Comment:
For the `render` submodule, you might be interested in [`cookie-factory`](https://crates.io/crates/cookie-factory), which produces data in the same way that `nom` consumes it 😋🍪🏭

Reply:
See, a proper schema library would only make you define the schema once... https://github.com/nuttycom/xenomorph